File: challenge/leet.gcc/easy/q414/q414.cpp
/**
 * 414. Third Maximum Number
 * https://leetcode.com/problems/third-maximum-number/description/
 */

#include <vector>
#include <set>
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    int partition(vector<int>& nums, int l, int r) {
        int p = nums[r];

        int left = l;
        for (int i = left; i <= r; i++) {
            if (nums[i] > p) {
                if (left < i)
                    swap(nums[i], nums[left]);
                left++;
            }
        }
        swap(nums[left], nums[r]);
        return left;
    }

    int kthMax(vector<int>& nums, int k, int l, int r) {
        int p = partition(nums, l, r);
        if (k - 1 == p - l) return nums[p];
        else if (k - 1 < p - l)
            return kthMax(nums, k, l, p - 1);
        else
            return kthMax(nums, k - 1 - (p - l), p + 1, r);
    }

    int thirdMax(vector<int>& nums) {
        set<int> s(nums.begin(), nums.end());
        nums = vector<int>(s.begin(), s.end());
        if (nums.size() == 1)
            return nums[0];
        else if (nums.size() == 2)
            return nums[0] > nums[1] ? nums[0] : nums[1];
        else
            return kthMax(nums, 3, 0, nums.size() - 1);
    }
};

class Solution2 {
public:

    int thirdMax(vector<int>& nums) {
        make_heap(nums.begin(), nums.end());
        set<int> s;

        int p;
        while (s.size() < 3 && nums.size() > 0) {
            p = nums.front();
            pop_heap(nums.begin(), nums.end());
            nums.pop_back();
            s.insert(p);
        }

        if (s.size() == 3)
            return p;
        else if (s.size() == 2) {
            int a = *s.begin(), b = *next(s.begin());
            return a > b ? a : b;
        }
        else if(s.size() == 1) {
            return *s.begin();
        }
        else
            return 0;
    }
};

int main() {
    Solution2 s;
    vector<int>v;

    v = vector<int>{ 3,2,1 };
    cout << "1?" << s.thirdMax(v) << endl;

    v = vector<int>{ 1,2 };
    cout << "2?" << s.thirdMax(v) << endl;

    v = vector<int>{ 1,1,2 };
    cout << "2?" << s.thirdMax(v) << endl;

    v = vector<int>{ 2,2,3,1 };
    cout << "1?" << s.thirdMax(v) << endl;

    v = vector<int>{ 1,4,2,3 };
    cout << "2?" << s.thirdMax(v) << endl;

    v = vector<int>{ 0, 1, 5, 7, 2 };
    cout << "2?" << s.thirdMax(v) << endl;
}

File: challenge/leet/easy/q414.py
'''
414. Third Maximum Number
https://leetcode.com/problems/third-maximum-number/
'''
import unittest
from heapq import heappush, heappop
from typing import List


class Solution:
    """
    110ms
    """
    def thirdMax(self, nums: List[int]) -> int:
        nums = list(set(nums))
        def partition(k: int, l: int, r: int) -> int:
            p = nums[r]
            for i in range(l, r):
                if nums[i] > p: # 1, 4, 2, 3 -> 4, 1, 2, 3 -> 4, 1, 2, 3
                    if l < i:
                        nums[l], nums[i] = nums[i], nums[l]
                    l += 1

            nums[l], nums[r] = nums[r], nums[l]
            return l

        def kthMax(k:int, l: int, r: int) -> int:
            p = partition(k, l, r)
            if p-l == k-1:
                return nums[p]
            elif p-l < k-1: # k in (p, r)
                return kthMax(k-1-p+l, p+1, r)
            else: # k in (l, p)
                return kthMax(k, l, p-1)

        ln = len(nums)
        if ln == 1:
            return nums[0]
        elif ln == 2:
            return max(nums)
        else:
            return kthMax(3, 0, ln - 1)


class Solution2:
    """
    67ms 13%
    """
    def thirdMax(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        elif len(nums) <= 2:
            return max(nums)
        else:
            s = set(nums)
            if len(s) == 1:
                return nums[0]
            elif len(s) <= 2:
                return max(s)
            nums = sorted(s)
            return nums[-3]


class Solution3:
    def thirdMax(self, nums: List[int]) -> int:
        if len(nums) < 3:
            return max(nums)

        hq = []
        for n in nums:
            heappush(hq, -n)

        s = set()
        while len(s) < 3 and len(hq) > 0:
            h = heappop(hq)
            s.add(h)

        return -h if len(s) == 3 else - min(s)


class MyTestCase(unittest.TestCase):
    def test_something(self):
        s = Solution3()
        self.assertEqual(1, s.thirdMax([1]))
        self.assertEqual(1, s.thirdMax([1, 0]))
        self.assertEqual(1, s.thirdMax([1, 2, 3]))
        self.assertEqual(1, s.thirdMax([1, 2, 2, 3]))
        self.assertEqual(1, s.thirdMax([1, 1, 1]))


if __name__ == '__main__':
    unittest.main()

